### PR TITLE
Ecc injection enable

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

File: src/mainboard/pcengines/apu2/mainboard.c [hunk 1]
```diff
@@ -250,6 +250,34 @@ static void mainboard_enable(struct device *dev)
...
+	// AGESA set bit 0, EccExclEn, in register D18F5x240 as 1.
+	// Range was incorrectly enabled to cover all memory
+	// for cases without UMA (no integrated graphics).
+	//
+	// In order to clear EccExclEn DRAM scrubber needs to be
+	// disabled temporarily by setting D18F3x88[DisDramScrub]
+	//
+	u32 val;
+	device_t D18F3 = dev_find_device(0x1022, 0x1583, NULL);
+	device_t D18F5 = dev_find_device(0x1022, 0x1585, NULL);
```
Comment (pietrushnic) [inline src/mainboard/pcengines/apu2/mainboard.c]: @krystian-hebel IMO variables names are not very descriptive. Do we really don't know what this device is?

File: src/mainboard/pcengines/apu2/mainboard.c [hunk 2]
```diff
@@ -250,6 +250,34 @@ static void mainboard_enable(struct device *dev)
...
 static void mainboard_final(void *chip_info)
 {
+	//
+	// AGESA set bit 0, EccExclEn, in register D18F5x240 as 1.
+	// Range was incorrectly enabled to cover all memory
+	// for cases without UMA (no integrated graphics).
+	//
+	// In order to clear EccExclEn DRAM scrubber needs to be
+	// disabled temporarily by setting D18F3x88[DisDramScrub]
+	//
```
Comment (pietrushnic) [inline src/mainboard/pcengines/apu2/mainboard.c]: @krystian-hebel does coreboot use this style of comments shouldn't it be:

```
/*
 * Foobar
 */
``` ?

File: src/mainboard/pcengines/apu2/mainboard.c [hunk 3]
```diff
@@ -250,6 +250,34 @@ static void mainboard_enable(struct device *dev)
...
+	u32 val;
+	device_t D18F3 = dev_find_device(0x1022, 0x1583, NULL);
+	device_t D18F5 = dev_find_device(0x1022, 0x1585, NULL);
+
+	val = pci_read_config32(D18F3, 0x88);
+	val |= (1 << 27);		// Disable DRAM ECC scrubbing
```
Comment (pietrushnic) [inline src/mainboard/pcengines/apu2/mainboard.c]: @krystian-hebel I'm not sure if inline comments are good in coreboot code.

File: src/mainboard/pcengines/apu2/mainboard.c [hunk 4]
```diff
@@ -250,6 +250,34 @@ static void mainboard_enable(struct device *dev)
...
+	pci_write_config32(D18F3, 0x88, val);
+
+	val = pci_read_config32(D18F3, 0xB8);
+	val &= 0xF000003F;		// Reserved bits, but injection doesn't work if set
```
Comment (pietrushnic) [inline src/mainboard/pcengines/apu2/mainboard.c]: @krystian-hebel probably you should add not that this was figured out by experimentation.

File: src/mainboard/pcengines/apu2/mainboard.c [hunk 5]
```diff
@@ -250,6 +250,34 @@ static void mainboard_enable(struct device *dev)
 
 static void mainboard_final(void *chip_info)
 {
+	//
+	// AGESA set bit 0, EccExclEn, in register D18F5x240 as 1.
+	// Range was incorrectly enabled to cover all memory
+	// for cases without UMA (no integrated graphics).
+	//
+	// In order to clear EccExclEn DRAM scrubber needs to be
+	// disabled temporarily by setting D18F3x88[DisDramScrub]
```
Comment (pietrushnic) [inline src/mainboard/pcengines/apu2/mainboard.c]: @krystian-hebel that description doesn't explain what is your high level motivation.

**Comment by miczyg1** [comment]
@krystian-hebel I agree with @pietrushnic . Resolve all requested changes and I can approve

File: src/mainboard/pcengines/apu2/mainboard.c (hunk 1 above)
Comment (krystian-hebel) [reply src/mainboard/pcengines/apu2/mainboard.c]: `Host bridge: Advanced Micro Devices, Inc. [AMD] Device 1583`, which isn't any more descriptive in my opinion.

Comment (miczyg1) [reply src/mainboard/pcengines/apu2/mainboard.c]: @pietrushnic I also have not found a name which could satisfy You. Looked in BKDG and other sources, but found only that these are all memory controller/northbridge devices

**Comment by miczyg1** [comment]
@pietrushnic do You have any more change requests? I have checked the change on all apus and it works with memtest Pro

File: src/mainboard/pcengines/apu2/mainboard.c (hunk 1 above)
Comment (pietrushnic) [reply src/mainboard/pcengines/apu2/mainboard.c]: @miczyg1 @krystian-hebel it looks fine to me, let's send that upstream and see what will happen.

**Pull request merged**